src/passes/function_passes/CheckPointPass.cpp

```diff
@@ -12,179 +12,220 @@
+  // Map each function name to the patchpoint IDs of the stackmap calls
   // in the function. This is used to ensure the runtime can associate the
   // stackmap calls inside each function with stackmap calls from the
-  // unoptimized version of the function
-  static std::map<StringRef, uint64_t> start_sm_id;
+  // unoptimized version of the function.
+  static map<StringRef, vector<uint64_t>> stackMaps;
 
   CheckPointPass() : FunctionPass(id) {}
 
+  virtual bool doInitialization(Module &mod) {
+    // Declare the guard failure function. This is necessary because the
+    // `__guard_failure` function is not defined in the current module.
+    LLVMContext &ctx = mod.getContext();
+    Type* i64 = Type::getInt64Ty(ctx);
+    FunctionType* signature = FunctionType::get(Type::getVoidTy(ctx),
+                                                i64, false);
+    Function *stackmap_func = Function::Create(
+        signature, Function::ExternalLinkage, "__guard_failure", &mod);
+    return true;
+  }
+
   virtual bool runOnFunction(Function &fun) {
-    StringRef fun_name = fun.getName();
-    // This pass currently only operates on two functions: `trace` and
-    // `__unopt_trace`. This is because it is supposed to demonstrate how
-    // the runtime can resume the execution of `trace` in `__unopt_trace`.
-    if (!fun_name.endswith(TRACE_FUN_NAME)) {
-      return false;
-    }
+    StringRef funName = fun.getName();
     outs() << "Running CheckPointPass on function: " << fun.getName() << '\n';
-
-    // Each function has a corresponding unoptimized version, which starts
-    // with the `__unopt_` prefix (the 'twin').
-    StringRef twin_fun = getTwinName(fun_name);
-
-    // The next stackmap ID to assign
-    uint64_t curr_sm_id = sm_id;
-    if (start_sm_id.find(twin_fun) != start_sm_id.end()) {
-      // If the 'twin' of this function has already been processed, index
-      // stackmap calls starting from ~ID, where ID is the identifier of the
-      // first stackmap call in the 'twin'
-      curr_sm_id = ~start_sm_id[twin_fun];
-    }
-
     Module *mod = fun.getParent();
-    DataLayout data_layout(mod);
-    // sm_id is the ID of the first stackmap call in this function
-    start_sm_id[fun_name] = sm_id;
-    uint64_t call_inst_count = 0;
-
     Type *i8ptr_t = PointerType::getUnqual(
         IntegerType::getInt8Ty(mod->getContext()));
-    // The callback to call when a guard fails
-    Constant* gf_handler_ptr = ConstantExpr::getBitCast(
+    // The callback to call when a guard fails.
+    Constant* guardHandler = ConstantExpr::getBitCast(
         mod->getFunction(GUARD_FUN_NAME), i8ptr_t);
     for (auto &bb : fun) {
       for (BasicBlock::iterator it = bb.begin(); it != bb.end(); ++it) {
-        if (isa<CallInst>(it) &&
-            cast<CallInst>(*it).getCalledFunction()->getName() == "putchar") {
+        if (funName.endswith("more_indirection") && isa<ReturnInst>(it)) {
```

Comment:
There's a hardcoded function name in here. We should at least put an `XXX`.

Reply:
I can do that, but the point is that a 'guard' is to going inserted at that particular point in the code. There is no real logic behind why I've chosen that particular condition.